### DISCUSSION TITLE
Ideas For Creating Blocks On-Demand

### DISCUSSION THREAD

**Discussion by zicklag** [paritytech/substrate#10926] [Q&A]
In my application I'm running a permissioned blockchain where I would like to get very quick confirmation and finalization of blocks, and it can be assumed that all nodes will be on 99% of the time.

So far, using Aura and Grandpa I've set up a block time of 3.5 seconds and tweaked Grandpa's voting rule so it will not wait 2 blocks to vote on a block, and that, in my tests with the local dev node, provides timing that I'm satisfied as far as how fast transactions are confirmed and finalized.

I was thinking, though. I don't really need to be creating blocks ever 3.5 seconds, if there are no transactions. Tendermint has a feature where you can configure it not to produce empty blocks and I'm wondering if there's a simple way to get substrate to do the same thing.

There will be a large proportion of the time that my blockchain is running where there will be no transactions, so producing blocks every 3.5 seconds seems very wasteful of server resources that could be put elsewhere if possible.

I know that I can create my own block production engine or maybe fork Aura, but I wanted to ask to see if anybody had any thoughts or ideas first.

**Comment by zicklag** [comment]
I just discovered [`sc_consensus_manual_seal`](https://docs.rs/sc-consensus-manual-seal/latest/sc_consensus_manual_seal/) which says it's only suitable for testing, but has a similar idea to what I'm imagining, of creating blocks when it receives transactions.

I think I want something like a combination of Aura and the manual seal. Here's what I'm imagining:

1. Don't create blocks if there are no transactions.
2. When you get a transaction, wait a couple seconds and include any other transactions you get within that time period in the next block.
3. Rotate block producer every block.

**Reply by xlc** [reply]
It is not going to work. What happen if the next block producer is down?

**Reply by zicklag** [reply]
:thinking: If the transactions come in and the current block producer doesn't produce a block in the given slot, then the next block producer will be selected and it will produce the block for the next slot, maybe?

I'll have to get deeper into the Aura implementation to see if that makes sense.

**Reply by xlc** [reply]
Then you are allowing multiple block producers per slot and therefore have fork problems to deal with. It ain't that simple.

**Reply by zicklag** [reply]
Maybe it's as simple as just using Aura and telling Aura not to produce a block if there are no transactions in it?

I could imagine that causing issues in some use-cases, for instance, when you want to have a more reliably updated timestamp, or other inherents that you want updated more often, but otherwise maybe that could work?

**Reply by zicklag** [reply]
Actually, I might be able to do that with a custom block importer that just ignores empty blocks, and then I wouldn't even have to modify Aura.

I'm going to try that!

**Reply by zicklag** [reply]
Ah, it looks like there's not a way for me to tell from the `BlockImportParams` what the transaction count was. I can count extrinsics, but because the timestamp would be an extrinsic, too, it doesn't tell me whether or not there were any _transactions_ if I'm understanding correctly.

I might need to fork Aura, and maybe submit a PR for an option to not produce blocks without transactions. I'll look more into that.

**Edit:** I might just wait for now and check this out later. I probably don't need it that badly, and I can come back to it when I need it.
